homeassistant/components/knx/config_flow.py

```diff
@@ -0,0 +1,254 @@
+        fields = {}
+
+        if user_input is None:
+            gateways = await scan_for_gateways()
+
+            if gateways:
+                supported_connection_types.insert(0, CONF_KNX_AUTOMATIC)
+                self._tunnels = [
+                    gateway for gateway in gateways if gateway.supports_tunnelling
+                ]
+
+            fields = {
+                vol.Required(CONF_KNX_CONNECTION_TYPE): vol.In(
+                    supported_connection_types
+                )
+            }
+
+        if user_input is not None:
+            connection_type = user_input[CONF_KNX_CONNECTION_TYPE]
+            if connection_type == CONF_KNX_AUTOMATIC:
+                return self.async_create_entry(
+                    title=CONF_KNX_AUTOMATIC, data=user_input
+                )
+
+            if connection_type == CONF_KNX_ROUTING:
+                return await self.async_step_routing()
+
+            if connection_type == CONF_KNX_TUNNELING and self._tunnels:
+                return await self.async_step_tunnel()
+
+            return await self.async_step_manual_tunnel()
+
+        return self.async_show_form(
+            step_id="type", data_schema=vol.Schema(fields), errors=errors
+        )
+
+    async def async_step_manual_tunnel(
+        self, user_input: dict | None = None
+    ) -> FlowResult:
+        """General setup."""
+        errors: dict = {}
+
+        if user_input is not None:
+            return self.async_create_entry(
+                title=user_input[CONF_HOST],
+                data={
+                    CONF_HOST: user_input[CONF_HOST],
+                    CONF_PORT: user_input[CONF_PORT],
+                    CONF_KNX_INDIVIDUAL_ADDRESS: user_input[
+                        CONF_KNX_INDIVIDUAL_ADDRESS
+                    ],
+                    ConnectionSchema.CONF_KNX_ROUTE_BACK: user_input[
+                        ConnectionSchema.CONF_KNX_ROUTE_BACK
+                    ],
+                    CONF_KNX_CONNECTION_TYPE: CONF_KNX_TUNNELING,
+                },
+            )
+
+        fields = {
+            vol.Required(CONF_HOST, default=self._gateway_ip): str,
+            vol.Required(CONF_PORT, default=self._gateway_port): vol.Coerce(int),
+            vol.Required(
+                CONF_KNX_INDIVIDUAL_ADDRESS, default=XKNX.DEFAULT_ADDRESS
+            ): str,
+            vol.Required(
+                ConnectionSchema.CONF_KNX_ROUTE_BACK, default=False
+            ): vol.Coerce(bool),
+        }
+
+        return self.async_show_form(
+            step_id="manual_tunnel", data_schema=vol.Schema(fields), errors=errors
+        )
+
+    async def async_step_tunnel(self, user_input: dict | None = None) -> FlowResult:
+        """Select a tunnel from a list. Will be skipped if the gateway scan was unsuccessful or if only one gateway was found."""
+        errors: dict = {}
+
+        if user_input is not None:
+            gateway: GatewayDescriptor = next(
+                filter(
```

Comment:
A generator expression is preferred over filter + lambda.

Reply:
Thanks, this indeed looks much cleaner